tripal/tripal
# Property Field supports multiple property types on the same page

# Bug Fix

### Issue #1398 


### Tripal Version: 4

## Description


Previously when there were two property fields on the same page, they would swap data. This is a huge bug as there will almost always be multiple property fields on the same page as each one represents a type of metadata!

This was happening because ChadoStorage was keeping track of chado records to be managed by chado table and not taking into account that we don't always want to merge all fields working on a single table into a single record! It was a hard bug to solve as the assumption went really deep in the chado storage class and we did need to support both cases: 

1. Multiple fields work together to describe a single chado record (e.g. organism genus, species, infraspecific type and infraspecific name fields all work together to describe a single organism record).
2. Multiple fields access the same table but do not want to have their values combined into a single record (e.g. two property fields where each one has all the information for a single record in the organismprop table).

## Solution

Allow fields to define table alias'. 
By default, the table alias is the same as the table name which makes this fix backwards compatible and ensures case A above still works.
Fields who want to keep their records separate can now indicate a table alias on a per StoragePropertyType basis.

Only data from StoragePropertyTypes referencing the same table alias will combine data to form a single chado record. 

## Additional Fixes

While adding automated testing around the ChadoStorage actions to ensure this works as designed and that no actions were broken by these changes, I also fixed the following smaller bugs:

 - Throws an error if store_id actions are used on a non-base table.
 - Throws an error if developers try to use table alias' with the store_id action since base tables do not support alias'
 - Read_value (join) actions were only working if there were other fields on the base table. This assumption has been removed.
 - Ensure that the read_value table alias' was handled differently for joins and base tables as it needed to be.

## Testing?




All the existing automated testing still passes showing this change is 100% backwards compatible.

I also added full automated testing of the various actions to ensure this addition of table alias' didn't break any of the actions and weed out incorrect assumptions that may have been made.

### Manually test content types in general

Manually test each of the content types ensuring that the default fields still work as expected. For example, create an organism without adding any new fields and ensure that the page saves and updates the data for each form element.

### Test the property field

1. Choose a content type supporting properties (e.g. Organism, Gene, Germplasm Accession, etc)
2. Navigate to Tripal > Page Structure > CONTENT TYPE > Manage Fields.
3. Add multiple chado property fields by repeating the following steps at least twice with different terms each time.
    a. Click on "Create new field"
    b. Select "Chado Property" under "Add a new field"
    c. Then enter a label (e.g. TestProp1) and click "Save and continue"
    d. Next choose the chado base table (e.g. if your using the gene content type then choose feature) and change the "Allowed number of values" to "unlimited". Then click "Save field settings"
    e. Finally set the term by scrolling to the bottom of the next form and entering something in the "Set the Term" form element. Examples include "Kingdom", "Phylum", "comment", "Subgroup". Really choose anything you like here. It just needs to be different for each property field.
4. Once you have multiple property fields added to the same content type, then go ahead and create content for that content type! (i.e. Admin > Tripal > Content > Add Tripal Content > CONTENT TYPE). Make sure to test multiple combinations of filling in fields, changing them, deleting them, etc and ensure the resulting page reflects those changes/values. Specifically, ensure that a value entered for one property fields is always associated with that field and never any of the others!

The following screenshots are showing one of my manual tests on the genetic marker content type with three properties (TestProp1, TestProp2 and TestPropSingle) where the last one is set to limited values and the first two are unlimited.
 - [Genetic Marker Insert Form](https://github.com/tripal/tripal/assets/1566301/09b637a2-24ec-4e81-8747-a67fb497a934)
 - [Page after submitting above form](https://github.com/tripal/tripal/assets/1566301/84b8dd8e-8348-4114-9fb1-ac3616b11d97)
 - [Update form with changes](https://github.com/tripal/tripal/assets/1566301/fc0da423-1b3d-4ee2-890c-bb6e7398d761)
 - [Page after submitting the updates](https://github.com/tripal/tripal/assets/1566301/bcaa1d01-0534-4227-a384-db39ee197a2b)
